File: utils_go/sign.go
package utils_go

import (
	v1 "proto/go"

	ec "github.com/bsv-blockchain/go-sdk/primitives/ec"
	hash "github.com/bsv-blockchain/go-sdk/primitives/hash"
	"google.golang.org/protobuf/proto"
)

// SignEnvelope calculates signature over deterministic marshal of the envelope
// with Signature fields cleared and fills Signature & SignatureAlgo.
func SignEnvelope(env *v1.Envelope, priv *ec.PrivateKey) error {
	// clone and clear sig
	clone := proto.Clone(env).(*v1.Envelope)
	clone.Signature = nil
	clone.SignatureAlgo = ""

	b, err := DeterministicMarshal(clone)
	if err != nil {
		return err
	}
	digest := hash.Sha256(b)

	sig, err := priv.Sign(digest)
	if err != nil {
		return err
	}
	der, err := sig.ToDER()
	if err != nil {
		return err
	}
	env.Signature = der
	env.SignatureAlgo = "ECDSA_SECP256K1_SHA256"
	return nil
}

File: utils_ts/sign.ts
// @ts-nocheck
/* eslint-disable */
import { deterministicMarshal } from "./encode";
import { Envelope } from "../ts/message";
import PrivateKey from "@bsv/sdk/primitives/PrivateKey";
import { sha256 } from "@bsv/sdk/primitives/Hash";
import { sign as brc77Sign } from "@bsv/sdk/messages/SignedMessage";
import { Buffer } from "buffer";

export function signEnvelope(env: Envelope, priv: PrivateKey): void {
  const clone: Envelope = { ...env, signature: Buffer.alloc(0), signatureAlgo: "" } as Envelope;
  const sigInput = deterministicMarshal(clone);
  const digest = sha256(sigInput);
  const sigBytes = brc77Sign(digest, priv, undefined);
  env.signature = Buffer.from(sigBytes);
  env.signatureAlgo = "ECDSA_SECP256K1_SHA256";
} 